### ISSUE TITLE
TF MirroredStrategy per-replica buffer corruption

### ISSUE THREAD

**Issue by ynikulin-tpx** [tensorflow/tensorflow#119013]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

binary

### TensorFlow version

2.19.0, also 2.14.1 and tf-nightly

### Custom code

Yes

### OS platform and distribution

Ubuntu 22.04

### Python version

3.12.11, also tested with 3.10.16 - bug was seen with both

### Current behavior?

We had reproducibility problems with our models. When all places in our code were checked, we started to suspect TF distribute internals, since strange deviation in predictions starts after distribute gathering. The code we provide reproduces the problem without external deps - note how the output which must be in [0, 1] after sigmoid gets huge values. 

Note that from time to time the script can pass without errors, need to launch multiple times.

We saw the corruption reproduced stably on 2 machines : one with NVidia RTX A5000 and the other with NVidia V100 16Gb PCI. On another machine with V100 32Gb SXM-2 we could not reproduce the bug in a few shots. 

### Standalone code to reproduce the issue

```shell
"""Minimal reproducer for TF MirroredStrategy per-replica buffer corruption.

Run on a host with ≥2 GPUs. The canary is the `scores` output, which is the
output of sigmoid and therefore must lie in [0, 1]. Any reading outside that
range means the per-replica handle for replica 1 (or higher) is pointing into
a recycled GPU buffer.
"""
import numpy as np
import tensorflow as tf

gpus = tf.config.list_physical_devices("GPU")
assert len(gpus) >= 2, f"need ≥2 GPUs, got {len(gpus)}"

strategy = tf.distribute.MirroredStrategy()
print(f"replicas: {strategy.num_replicas_in_sync}")

# Match the orders of magnitude from the failing real workload
B_PER_REPLICA = 1
H, W = 2304, 1792
NUM_ANCHORS = 192_780
NUM_ENSEMBLE = 2          # two "model versions" — like test_folds=[0, 2]

with strategy.scope():
    # Heavy-ish CNN to create allocator pressure; doesn't need to be a real model.
    # NOTE: Input layer added so Keras builds eagerly under strategy.scope() — without
    # it, Keras 3 (TF 2.16+) defers the build until first call, which races inside
    # strategy.run and crashes with "Input to reshape is a tensor with 0 values".
    backbones = [
        tf.keras.Sequential([
            tf.keras.layers.Input(shape=(H, W, 1)),
            tf.keras.layers.Conv2D(16, 3, strides=4, padding="same", activation="relu"),
            tf.keras.layers.Conv2D(32, 3, strides=4, padding="same", activation="relu"),
            tf.keras.layers.Conv2D(64, 3, strides=4, padding="same", activation="relu"),
            tf.keras.layers.GlobalAveragePooling2D(),
        ])
        for _ in range(NUM_ENSEMBLE)
    ]
    # Dense heads also need eager build — give them their input dim explicitly.
    box_heads = [tf.keras.layers.Dense(NUM_ANCHORS * 4) for _ in range(NUM_ENSEMBLE)]
    score_heads = [tf.keras.layers.Dense(NUM_ANCHORS) for _ in range(NUM_ENSEMBLE)]
    for h in box_heads + score_heads:
        h.build((None, 64))
    # A large constant per replica — mimics anchors_for_batch_tf that we suspect
    # gets recycled into the scores output buffer on GPU 1
    anchors_const = tf.constant(np.random.randn(1, NUM_ANCHORS, 4).astype(np.float32) * 500)

def forward(images):
    box_list, score_list = [], []
    bs = tf.shape(images)[0]
    anchors_b = tf.tile(anchors_const, (bs, 1, 1))   # the suspect intermediate
    for bb, bh, sh in zip(backbones, box_heads, score_heads):
        feat = bb(images)                            # (B, 64)
        b = tf.reshape(bh(feat), (-1, NUM_ANCHORS, 4)) + anchors_b
        s = tf.nn.sigmoid(sh(feat))                  # (B, NUM_ANCHORS), in [0, 1]
        box_list.append(b)
        score_list.append(s)
    boxes = tf.reduce_mean(box_list, axis=0)
    scores = tf.reduce_mean(score_list, axis=0)
    # Return dict of multiple float32 tensors of different sizes — key pattern
    return {"boxes": boxes, "scores": scores}

@tf.function
def step(images):
    per_replica = strategy.run(forward, args=(images,))
    out = {}
    for name, val in per_replica.items():
        out[name] = tf.concat(strategy.experimental_local_results(val), axis=0)
    return out

global_b = B_PER_REPLICA * strategy.num_replicas_in_sync
input_shape = (global_b, H, W, 1)
n_iters = 100

for i in range(n_iters):
    images = tf.random.normal(input_shape, dtype=tf.float32)
    out = step(images)
    s = out["scores"].numpy()
    if s.max() > 1.0 + 1e-3 or s.min() < -1e-3:
        print(f"\n!!! CORRUPTION at iter {i} !!!")
        print(f"    scores shape={s.shape}, range=[{s.min():.6g}, {s.max():.6g}]")
        for r in range(s.shape[0]):
            print(f"    replica/row {r}: [{s[r].min():.6g}, {s[r].max():.6g}]")
        break
    if i % 25 == 0:
        print(f"iter {i:4d}: ok, scores in [{s.min():.4g}, {s.max():.4g}]")
else:
    print(f"\nNo corruption in {n_iters} iterations.")
```

### Relevant log output

```shell
(loki) tpx-docker /media/gaia/ssd/homedir/ynikulin/tpx-repos/A/tpx-r/mammodiag/tests/tf/models/hn > CUDA_VISIBLE_DEVICES=8,9 python ./tmp.py 
2026-05-15 12:39:42.198613: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2026-05-15 12:39:42.216815: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:467] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
E0000 00:00:1778848782.239416  113462 cuda_dnn.cc:8579] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1778848782.245243  113462 cuda_blas.cc:1407] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
W0000 00:00:1778848782.264056  113462 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1778848782.264099  113462 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1778848782.264104  113462 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1778848782.264108  113462 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
2026-05-15 12:39:42.269427: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F AVX512_VNNI FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2026-05-15 12:39:50.532021: W tensorflow/core/common_runtime/gpu/gpu_bfc_allocator.cc:47] Overriding orig_value setting because the TF_FORCE_GPU_ALLOW_GROWTH environment variable is set. Original config value was 0.
I0000 00:00:1778848790.533013  113462 gpu_device.cc:2019] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 22364 MB memory:  -> device: 0, name: NVIDIA RTX A5000, pci bus id: 0000:d5:00.0, compute capability: 8.6
2026-05-15 12:39:50.533824: W tensorflow/core/common_runtime/gpu/gpu_bfc_allocator.cc:47] Overriding orig_value setting because the TF_FORCE_GPU_ALLOW_GROWTH environment variable is set. Original config value was 0.
I0000 00:00:1778848790.534836  113462 gpu_device.cc:2019] Created device /job:localhost/replica:0/task:0/device:GPU:1 with 22364 MB memory:  -> device: 1, name: NVIDIA RTX A5000, pci bus id: 0000:d6:00.0, compute capability: 8.6
replicas: 2
I0000 00:00:1778848792.864046  113582 cuda_dnn.cc:529] Loaded cuDNN version 90300
I0000 00:00:1778848792.864553  113581 cuda_dnn.cc:529] Loaded cuDNN version 90300
iter    0: ok, scores in [nan, nan]

!!! CORRUPTION at iter 1 !!!
    scores shape=(4, 192780), range=[-2363.82, 2265.32]
    replica/row 0: [0.499005, 0.500996]
    replica/row 1: [0.499011, 0.500994]
    replica/row 2: [-2165.28, 2265.32]
    replica/row 3: [-2363.82, 2247.48]
```

**Comment by bharatjetti** [comment]
Hi @ynikulin-tpx, Thanks for reporting this and sharing the details.
We’ve received the issue and are taking a look. We’ll follow up here once we have more information or if we need any additional inputs.
Appreciate your patience.

**Comment by Rajith-14** [comment]
Hi @ynikulin-tpx, thanks for the detailed report and the clean reproducer!

This looks like a GPU buffer aliasing bug in `BFCAllocator` interacting with `MirroredStrategy`. The pattern — replica 0/1 correct, replica 2/3 corrupted with values that match the magnitude of `anchors_b` (~500 scale) — strongly suggests the allocator is reusing the freed `anchors_b` intermediate buffer as the output buffer for `scores` on higher replicas.

The fact that it doesn't reproduce on V100 32GB SXM-2 but does on A5000 and V100 16GB PCI supports this: on the larger-memory machine, the allocator finds a fresh block rather than recycling the freed one.

**Workarounds to try while this is being investigated:**

1. Break the aliasing with `tf.identity` on the suspect intermediate:
```python
anchors_b = tf.identity(tf.tile(anchors_const, (bs, 1, 1)))
```

2. Switch to the async CUDA allocator:
```bash
export TF_GPU_ALLOCATOR=cuda_malloc_async
```

3. Remove `@tf.function` to run eagerly — slower but avoids graph-mode buffer reuse entirely.

Could you confirm which TF version and CUDA version you're on, and whether `TF_GPU_ALLOCATOR=cuda_malloc_async` changes the behavior? That would help narrow down whether this is in `BFCAllocator` specifically or deeper in XLA buffer assignment.

**Comment by ynikulin-tpx** [comment]
So I tried TF versions 2.19.0, also 2.14.1 and also tf-nightly (the day before I posted the 1st message).

`TF_GPU_ALLOCATOR=cuda_malloc_async` alone doesn't change the behavior. However, thanks to your hints, the combo of 3 things (no `@tf.function`, gather per-replica results on CPU and build anchors per replica) helped us to bypass the problem - I join here the updated version of the script which stably passes on machines where the first version of the script failed regularly.



[mirrored_strategy_p2p_fix.py](https://github.com/user-attachments/files/28789577/mirrored_strategy_p2p_fix.py)

**Comment by ankitxvx** [comment]
Hi @Venkat6871, 

Hypothesis: The intermittent buffer corruption during MirroredStrategy aggregation strongly indicates an allocator/stream-synchronization bug in gpu_bfc_allocator.cc or the device-to-device copy path. If pooled GPU memory blocks are recycled before async cross-stream transfers complete, subsequent allocations overwrite live tensor memory.

Plan:

Verify if wrapping per-replica results in tf.identity or forcing host-sync eliminates the issue (confirming an stream/lifetime race).

Trace the MirroredStrategy gather and GPU BFC Allocator code paths to ensure device events (cudaEventRecord) properly block memory chunk recycling.

Add a multi-GPU regression test reproducing the concurrent allocation stress.

Could you please assign this issue to me?
